File: DSA/Assignment_6/Q1.java
/*Name-Anish More
 Registration Number-2241013365
 Name Of the program-Static Implementation (Array Implementation)
A stack is implemented statically by using an array of size MAX to hold stack elements and an
integer top storing the position of top of the stack. The stack elements can be integers, characters,
strings or user defined data types.
The operations to be performed on a stack are
public static int push(int S[],int top) – adding an element x to the stack S
public static int pop(int S[],int top)– deletes and returns the top element from the stack
S
public static void display(int S[],int top)- display all the elements of Stack S
public static boolean isEmpty(int top) – check if the stack is empty
public static boolean isFull(int top) – check if the stack is full when top equals MAX -1
Write a menu driven Java Program using class, methods and array, to construct a Stack and
implement the above five operations.*/
import java.util.Scanner;
public class Q1 
{
	public static final int MAX=10;
	
	
	public static boolean isEmpty(int top){
		if(top==-1)
			return true;
		return false;
	}
	
	
	public static boolean isFull(int top){
		if(top==MAX-1)
			return true;
		return false;
	}
	
	
	public static int push(int S[],int top) 
	{
		if(isFull(top))
		{
			System.out.println("Stack Overflow");
			return top;
		}
		Scanner sc=new Scanner(System.in);
		System.out.print("Enter the element to insert:");
		S[++top]=sc.nextInt();
		return top;
	}
	
	
	public static int pop(int S[],int top)
	{
		if(isEmpty(top))
		{
			System.out.println("Stack Underflow");
			return top;
		}
		System.out.println("Deleted item is: "+S[top]);
		return --top;
	}
	
	
	public static void display(int S[],int top) 
	{
		if(isEmpty(top))
		{
			System.out.println("Stack is Empty");
		}
		else {
			System.out.println("Stack Elements are: ");
			for(int i=top;i>=0;i--)
			{
				System.out.println(S[i]);
			}
		}
	}
	
	
	public static void main(String[] args) {
	Scanner sc=new Scanner(System.in);
	int stack[]=new int[MAX];
	int top=-1;
	while(true)
	 {
		System.out.println("***MENU***");
		System.out.println("0: Exit");
		System.out.println("1: Push");
		System.out.println("2: Pop");
		System.out.println("3: Display");
		System.out.println("Enter your choice");
		int choice=sc.nextInt();
		switch(choice)
		{
			case 0:
				System.exit(0);
			case 1:
				top=push(stack,top);
				break;
			case 2:
				top=pop(stack,top);
				break;
			case 3:
				display(stack,top);
				break;
			default:
				System.out.println("Invalid choice");
			}
	 	}
	}
}

File: PPWC/Stack/Q1.c
#include <stdio.h>
#include <stdlib.h>
struct Stack{
    int top;
    int size;
    int *arr;
};

void push(struct Stack *s, int data);
int pop(struct Stack *s);
int peek(struct Stack s);
int isEmpty(struct Stack s);
int isFull(struct Stack s);
int count(struct Stack s);
void display(struct Stack s);
 

int main(void){
    struct Stack s;
    printf("Enter the size of the stack: ");
    scanf("%d",&s.size);
    s.arr = (int *)malloc(s.size*sizeof(int));
    s.top = -1;
    int ch=1,val;
    while (ch){
        printf("1. Push\n2. Pop\n3. Peek\n4. Count\n5. Display\n6. Exit\n");
        int choice;
        printf("Enter your choice: ");
        scanf("%d",&choice);
        switch(choice){
            case 1:
                printf("Enter the element to push: ");
                int data;
                scanf("%d",&data);
                push(&s,data);
                break;
            case 2:
                val = pop(&s);
                if(val!=-1)
                    printf("Popped element is: %d\n",val);
                break;
            case 3:
                printf("Top element is: %d\n",peek(s));
                break;
            case 4:
                printf("Number of elements in the stack is: %d\n",count(s));
                break;
            case 5:
                display(s);
                break;
            case 6:
                ch=0;
                break;
            default:
                printf("Invalid choice\n");
        }
    }
    return 0;
}

int isEmpty(struct Stack s){
    if(s.top==-1){
        return 1;
    }
    return 0;
}

int isFull(struct Stack s){
    if(s.top==s.size-1){
        return 1;
    }
    return 0;
}

void push(struct Stack *s, int data){
    if(isFull(*s)){
        printf("Stack Overflow\n");
    }
    else{
        (*s).top++;
        (*s).arr[(*s).top] = data;
    }
}

int pop(struct Stack *s){
    if(isEmpty(*s)){
        printf("Stack Underflow\n");
        return -1;
    }
    else{
        int val = (*s).arr[(*s).top];
        (*s).top--;
        return val;
    }
}

int peek(struct Stack s){
    if(isEmpty(s)){
        printf("Stack Underflow\n");
        return -1;
    }
    else{
        return s.arr[s.top];
    }
}

int count(struct Stack s){
    return s.top+1;
}

void display(struct Stack s){
    if(isEmpty(s)){
        printf("Stack is empty\n");
    }
    else{
        for(int i=s.top;i>=0;i--){
            printf("%d ",s.arr[i]);
        }
        printf("\n");
    }
}

// Enter the size of the stack: 5
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 1
// Enter the element to push: 1
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 1
// Enter the element to push: 2
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 1
// Enter the element to push: 3
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 1
// Enter the element to push: 4
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 1
// Enter the element to push: 5
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 1
// Enter the element to push: 6
// Stack Overflow
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 5
// 5 4 3 2 1
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 4
// Number of elements in the stack is: 5
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 3
// Top element is: 5
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 2
// Popped element is: 5
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 2
// Popped element is: 4
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 2
// Popped element is: 3
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 2
// Popped element is: 2
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 2
// Popped element is: 1
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 2
// Stack Underflow
// 1. Push
// 2. Pop
// 3. Peek
// 4. Count
// 5. Display
// 6. Exit
// Enter your choice: 6